### PR TITLE
Remove some unnecessary checks and dead code from Linq.Expressions

### PR DESCRIPTION
**Pull request by JonHanna**
Remove dead code from KeyedQueue helper class.

Replace run-time check on impossible condition into assert.

ReadOnlyCollection is tested for condition of count < 0 though getting here would pass through paths that would either fail or produce an empty collection en route.

Include tests attepting all possible routes for getting such a bogus collection into an expression.

Turn run-time check on array creation expression having array type with assert, both at the point where the run-time check was, and at the point where the expressions are given that type.

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @VSadov

**Comment by JonHanna** [comment]
Test Innerloop OpenSUSE13.2 Release Build and Test please

**Comment by JonHanna** [comment]
Test Innerloop OSX Release Build and Test
Test Innerloop OpenSUSE13.2 Release Build and Test

File: src/System.Linq.Expressions/tests/Array/ArrayBoundsTests.cs [hunk 1]
```diff
@@ -1339,6 +1420,15 @@ public static void CheckUShortArrayWithUShortSize()
 
         #endregion
 
+        [Fact]
+        public static void ThrowOnNegativeSizedCollection()
+        {
+            // This is an obscure case, and it doesn't much mather what is thrown, as long as is thrown before such
+            // an edge case could cause more obscure damage. A class derived from ReadOnlyCollection is used to catch
```
Comment (VSadov) [inline src/System.Linq.Expressions/tests/Array/ArrayBoundsTests.cs]: mather?

**Comment by VSadov** [comment]
LGTM

**Comment by JonHanna** [comment]
Typo fixed, commits squashed.

**Comment by JonHanna** [comment]
Test Innerloop OSX Debug Build and Test please

**Comment by JonHanna** [comment]
Have added the removal of `LogInstruction` to this. The class is completely unused, apparently was a helper for the development process.

**Comment by JonHanna** [comment]
Test Innerloop CentOS7.1 Debug Build and Test please

**Pull request merged**